lifecycle/services/annotate.go

```diff
@@ -0,0 +1,174 @@
+	RepoKey   string `json:"repoKey"`
+}
+
+type AnnotateOperationParams struct {
+	RbTag          RbAnnotationTag
+	RbProps        RbAnnotationProps
+	RbDetails      ReleaseBundleDetails
+	QueryParams    CommonOptionalQueryParams
+	PropertyParams CommonPropParams
+}
+
+type RbAnnotationTag struct {
+	Tag string `json:"tag,omitempty"`
+}
+
+type RbAnnotationProps struct {
+	Properties map[string][]string `json:"properties,omitempty"`
+}
+
+func (rbs *ReleaseBundlesService) AnnotateReleaseBundle(rbDetails ReleaseBundleDetails, artUrl string, params AnnotateOperationParams) error {
+	return rbs.annotateReleaseBundle(params, GetReleaseBundleSetTagApi(rbDetails), GetReleaseBundleAddPropsApi(), artUrl)
+}
+
+func GetReleaseBundleSetTagApi(rbDetails ReleaseBundleDetails) string {
+	return path.Join(releaseBundleBaseApi, records, rbDetails.ReleaseBundleName, rbDetails.ReleaseBundleVersion, tag)
+}
+
+func GetReleaseBundleAddPropsApi() string {
+	return propertiesBaseApi
+}
+
+func (rbs *ReleaseBundlesService) setTag(params AnnotateOperationParams, api string, httpClientsDetails httputils.HttpClientDetails) error {
+	projParam := distribution.GetProjectQueryParam(params.QueryParams.ProjectKey)
+	tagContent, err := json.Marshal(params.RbTag)
+	if err != nil {
+		return errorutils.CheckError(err)
+	}
+
+	setTagFullUrl, err := utils.BuildUrl(rbs.GetLifecycleDetails().GetUrl(), api, projParam)
+	if err != nil {
+		return err
+	}
+
+	resp, body, err := rbs.client.SendPut(setTagFullUrl, tagContent, &httpClientsDetails)
+	if err != nil {
+		return err
+	}
+
+	if err = errorutils.CheckResponseStatusWithBody(resp, body, http.StatusOK); err != nil {
+		return err
+	}
+	return err
+}
+
+func (rbs *ReleaseBundlesService) setProps(params AnnotateOperationParams, api, artUrl string, httpClientsDetails httputils.HttpClientDetails) error {
+
+	propsContent, err := json.Marshal(params.RbProps)
+	if err != nil {
+		return errorutils.CheckError(err)
+	}
+
+	var initialStruct InitialProperties
+	err = json.Unmarshal(propsContent, &initialStruct)
+	if err != nil {
+		return errorutils.CheckError(err)
+	}
+
+	var propertyBodies []PropertyBody
+
+	propsFullUrl, err := utils.BuildUrl(artUrl, api, GetPropQueryParam(params.PropertyParams))
+	// Iterate through the properties map to create PropertyBody instances
+	for key, values := range initialStruct.Properties {
+		propertyBody := PropertyBody{
+			MultiValue:     false,
+			Property:       Property{Name: key},
+			SelectedValues: values,
+		}
+
+		propertyBodies = append(propertyBodies, propertyBody)
+	}
+
+	for _, pb := range propertyBodies {
+		content2, err2 := json.Marshal(pb)
+		if err2 != nil {
+			return errorutils.CheckError(err2)
+		}
+		resp2, body2, err2 := rbs.client.SendPost(propsFullUrl, content2, &httpClientsDetails)
+		if err2 != nil {
+			return err2
+		}
+		if err = errorutils.CheckResponseStatusWithBody(resp2, body2, http.StatusCreated, http.StatusOK); err != nil {
+			return err
+		}
+	}
```

Comment:
Couple of questions/comments here:
- there are 2 different for loops can this be changed to single for loop ?
- don't we have an API for taking multiple property bodies?
- can the variables be named with what it stores instead of content2, body2, resp2?
- reuse err instead of creating multiple err varaibles to ensure there is no shadowing.